pkg/kgo/source.go

```diff
@@ -39,6 +41,35 @@ func (p recordsPool) put(r *Record) {
 	p.p.Put(r)
 }
 
+// rcBuffer is a reference counted buffer.
+//
+// The internal buffer will be sent back to the pool after calling release
+// when the ref count reaches 0.
+type rcBuffer[T any] struct {
+	refCount atomic.Int32
+	buffer   []T
+	pool     *pool.BucketedPool[T]
+}
+
+func newRCBuffer[T any](buffer []T, pool *pool.BucketedPool[T]) *rcBuffer[T] {
+	return &rcBuffer[T]{buffer: buffer, pool: pool}
+}
+
+func (b *rcBuffer[T]) add(delta int32) {
+	b.refCount.Add(delta)
+}
+
+func (b *rcBuffer[T]) release() {
+	if b.refCount.Add(-1) == 0 {
+		b.pool.Put(b.buffer)
+		b.buffer = nil
+		return
+	}
+	if b.refCount.Load() < 0 {
+		panic("rcBuffer released too many times")
+	}
+}
```

Comment:
I don't quite understand the motivation for the method set on `rcBuffer`. Why have a `delta` parameter on `add` when this may only ever be called with an argument of `1` exactly once or else `release()` will panic?

Reply:
Good observation. The method was originally written with the idea of being called only once, but given the current usage, it would be more convenient to have the opposite equivalent to `release`.

_addressed @ d507d35ee3471178b13f6dddde59ea6fbc8f1fd7_